src/clp_ffi_js/ir/StructuredIrUnitHandler.cpp

```diff
@@ -0,0 +1,145 @@
+#include <cctype>
+#include <cstddef>
+#include <iterator>
+#include <memory>
+#include <string>
+#include <string_view>
+#include <type_utils.hpp>
+#include <utility>
+
+#include <clp/ffi/ir_stream/decoding_methods.hpp>
+#include <clp/ffi/KeyValuePairLogEvent.hpp>
+#include <clp/ffi/SchemaTree.hpp>
+#include <clp/ffi/Value.hpp>
+#include <emscripten/val.h>
+#include <spdlog/spdlog.h>
+
+#include <clp_ffi_js/constants.hpp>
+#include <clp_ffi_js/ir/LogEventWithFilterData.hpp>
+
+namespace clp_ffi_js::ir {
+
+namespace {
+/**
+ * Parses the `LogLevel` from an input string.
+ * @param str
+ * @return
+ */
+auto parse_log_level(std::string_view str) -> LogLevel;
+
+auto parse_log_level(std::string_view str) -> LogLevel {
+    LogLevel log_level{LogLevel::NONE};
+
+    // Convert the string to uppercase.
+    std::string log_level_name_upper_case{str};
+    std::ranges::transform(
+            log_level_name_upper_case.begin(),
+            log_level_name_upper_case.end(),
+            log_level_name_upper_case.begin(),
+            [](unsigned char c) { return std::toupper(c); }
+    );
+
+    auto const* it = std::ranges::find(
+            cLogLevelNames.begin() + clp::enum_to_underlying_type(cValidLogLevelsBeginIdx),
+            cLogLevelNames.end(),
+            log_level_name_upper_case
+    );
+
+    if (it == cLogLevelNames.end()) {
+        return log_level;
+    }
+
+    log_level = static_cast<LogLevel>(std::distance(cLogLevelNames.begin(), it));
+    return log_level;
+}
+}  // namespace
+
+auto StructuredIrUnitHandler::handle_schema_tree_node_insertion(
+        clp::ffi::SchemaTree::NodeLocator schema_tree_node_locator
+) -> clp::ffi::ir_stream::IRErrorCode {
+    ++m_current_node_id;
+
+    auto const& key_name{schema_tree_node_locator.get_key_name()};
+    if (m_log_level_key == key_name) {
+        m_log_level_node_id.emplace(m_current_node_id);
+    } else if (m_timestamp_key == key_name) {
+        m_timestamp_node_id.emplace(m_current_node_id);
+    }
+
+    return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+}
+
+auto StructuredIrUnitHandler::handle_log_event(StructuredLogEvent&& log_event
+) -> clp::ffi::ir_stream::IRErrorCode {
+    auto const& id_value_pairs{log_event.get_node_id_value_pairs()};
+    auto const timestamp = get_timestamp(id_value_pairs);
+    auto const log_level = get_log_level(id_value_pairs);
+
+    m_deserialized_log_events->emplace_back(std::move(log_event), log_level, timestamp);
+
+    return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+}
+
+auto StructuredIrUnitHandler::get_log_level(
+        StructuredLogEvent::NodeIdValuePairs const& id_value_pairs
+) const -> LogLevel {
+    LogLevel log_level{LogLevel::NONE};
+
+    if (false == m_log_level_node_id.has_value()) {
+        return log_level;
+    }
+
+    auto const& log_level_pair{id_value_pairs.at(m_log_level_node_id.value())};
+
+    if (false == log_level_pair.has_value()) {
+        return log_level;
+    }
```

Comment:
Would this look cleaner?
```suggestion
    if (false == m_log_level_node_id.has_value()) {
        return LogLevel::NONE;
    }

    auto const& log_level_pair{id_value_pairs.at(m_log_level_node_id.value())};
    if (false == log_level_pair.has_value()) {
        return LogLevel::NONE;
    }
```